### ISSUE TITLE
profile-credentials-provider does not support assumed roles

### ISSUE THREAD

**Issue by krestivo-kdinfotech** [babashka/pod-babashka-aws#56]
```
[foo]
aws_access_key_id=redacted
aws_secret_access_key=redacted
region=us-west-2
```
I can do babashka pods aws functions using the above profile.

```
[assumed]
role_arn = arn:aws:iam::671645499998:role/OrganizationAccountAccessRole
source_profile= foo
region=us-west-2
```
Does not work however.

It gives me:

```
{:cognitect.anomalies/category :cognitect.anomalies/fault,
 :cognitect.anomalies/message
 "Unable to fetch credentials. See log for more details."}
```

There is no "log" that I can tell from which I can see more details, but it's clear that either the Babashka AWS pod or the underlying Cognitect AWS API library does not support automatic STS assume role as for example boto3 and the AWS CLI written on top of it does.

I can hack around this by writing my own STS utility functions, but just FYI this will confuse people I'm sure.

**Comment by stevebuik** [comment]
same for basic-credentials-provider blocked me. 

my workaround is to do assume-role in shell and set env vars using jet. works well but not a nice as testing via repl

**Comment by jeroenvandijk** [comment]
@krestivo-kdinfotech These options you are referring to seem to be relatively new `~/.aws/credentials` format ([link](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html))

Cognitect AWS API library [only supports a very small part](https://github.com/cognitect-labs/aws-api/blob/master/src/cognitect/aws/credentials.clj#L240-L242) of the [v1 credentials file](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html)

We have decided to support the `credential_process` ([for background see](https://github.com/cognitect-labs/aws-api/issues/73))

With `credential_process` you can do basically anything you want with credentials ([documentation](https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes), [an example](https://gist.github.com/jeroenvandijk/ace7432be94d083e63729ac313a0b78f))

**Comment by stevebuik** [comment]
Thanks for the link but the example is very complex. That will keep me staying with my shell/jet workaround because it's much simpler.

sts assume-role is an AWS security best practice so maybe it's worth considering this for built-in support instead of needing a custom process?

**Comment by jeroenvandijk** [comment]
@stevebuik Maybe you can describe how you would like to see the `sts assume-role` to be built-in?

**Comment by stevebuik** [comment]
happy to. it would take 2 changes:

1. basic-credentials-provider needs a new passthrough arg for aws session token. this is required for assumed roles
2. a new credentials fn that calls assume role and load the results into a basic provider. something like this..
```clojure
(defn assumed-role-creds-provider
  "assume an AWS role and return a credentials-provider that will use the new keys to authenticate requests"
  ; https://github.com/babashka/pod-babashka-aws#credentials
  [{:keys [region role-arn]}]
  (let [sts-client (aws/client {:api    :sts
                                :region region})
        {:keys [Credentials] :as response}
        (aws/invoke sts-client
                    {:op      :AssumeRole
                     :request {:RoleArn         role-arn
                               :RoleSessionName "Hotplate-CI-deploy"}})]
    (if-let [k (:AccessKeyId Credentials)]
      (credentials/basic-credentials-provider
        {:access-key-id     k
         :secret-access-key (:SecretAccessKey Credentials)})
      (throw (ex-info "failed to assume role" response)))))
```

**Comment by jeroenvandijk** [comment]
@borkdude Do you think we could add support for the `CredentialProvider` protocol now that Babashka supports `reify` and `defprotocol`? 

None of the above described functionality is available in Cognitect's AWS API, however this lib supports this protocol making more customisations possible. If we want to prevent deviating too much away from the source lib, and spending more time on support probably, I think supporting this protocol would help. We could then refer to the source lib instead.

Here is an example how of how the above sts functionality (described by @stevebuik)  would work with the protocol: [link](https://github.com/cognitect-labs/aws-api/blob/master/examples/assume_role_example.clj#L50-L63)

**Comment by borkdude** [comment]
@jeroenvandijk Bb supports protocols and reify on protocols, but we're dealing with serialized objects from bb <-> pod which might complicate things since you can't serialize functions?

Doing this with https://github.com/grzm/awyeah-api might be easier - this is basically AWS API rewritten for babashka, from source.

**Comment by jeroenvandijk** [comment]
> Bb supports protocols and reify on protocols, but we're dealing with serialized objects from bb <-> pod which might complicate things since you can't serialize functions?

Ah yeah, forgot about this part. Sounds like it would indeed be complicated.

Maybe extending `basic-credentials-provider` with the `:session-token` field like @stevebuik suggested would give the required flexibility. Maybe better to give it a different name to avoid confusion with Cognitect's lib.

I didn't try https://github.com/grzm/awyeah-api yet, but that might be a good option too.

**Comment by borkdude** [comment]
Yeah, I think we can support some non-standard things as long as they have a clear name and some docs.
